### PR TITLE
119 feature add object bbox generation and event to object mapping for overlay payload

### LINKED ISSUE #119: [Feature] Add object bbox generation and event-to-object mapping for overlay payload
### Goal

Implement a lightweight backend/CV layer that produces object bounding boxes for action events and attaches only event-relevant boxes to the shared `ActionEvent` / `EventPayload` payload.

The implementation should use a pretrained object detector without training a new detector model. The detector should provide candidate object detections such as `person`, `car`, and `motorcycle`. These detections should then be filtered according to the action label predicted by the action-recognition model.

The action class predicted by the ML model is stored in `ActionEvent.label`, for example `car_drops_off_person`.

Current action labels:

```yaml
class_labels:
  - car_drops_off_person
  - car_makes_u_turn
  - motorcycle_makes_u_turn
  - motorcycle_turns_right
  - person_sits_down
```

The goal is to prepare real backend-generated bbox data for the frontend overlay using the optional `bboxes` contract introduced in the WebSocket/EventPayload work. The frontend should not rely on `mockCoordinates.json` as the main demo path. Bounding boxes should come from the backend event pipeline when they are available.

This issue should also integrate bbox enrichment with the shared frame-to-event pipeline from the final sprint work, so MP4, RTSP, and browser-camera flows can expose bounding boxes through the same event contract.

### Scope

In scope:

* Add a lightweight object detection integration using a pretrained detector, e.g. YOLO or equivalent.
* Do not train a new detector model.
* Detect only object classes relevant for the current action labels (check #130), for example:

  * `person`
  * `car`
  * `motorcycle`
* Add mapping from `ActionEvent.label` to allowed object classes, for example:

  * `car_drops_off_person` -> `car`, `person`
  * `car_makes_u_turn` -> `car`
  * `motorcycle_makes_u_turn` -> `motorcycle`
  * `motorcycle_turns_right` -> `motorcycle`
  * `person_sits_down` -> `person`
* Attach filtered object detections to `ActionEvent.bboxes` using the contract from #115.
* Support multiple bboxes per event.
* Keep bboxes optional: events without object detections must still be valid and backward-compatible.
* Include bbox metadata required by the frontend:

  * `box_format`, e.g. `xyxy`
  * `coordinate_space`, e.g. `source_pixels` or `normalized`
  * `frame_index`
  * `source_width`
  * `source_height`
  * `x_min`, `y_min`, `x_max`, `y_max`
  * `label`
  * `confidence`
* Use a representative frame for object detection, for example the middle or end frame of the inference window.
* Integrate bbox enrichment through the shared event pipeline or enrichment hook from the final sprint shared runtime work.
* Ensure the same bbox contract can be used by MP4, RTSP, and browser-camera event flows.
* Verify that the main dashboard no longer depends on `mockCoordinates.json` for overlay boxes.
* Provide safe fallback behavior, for example an empty bbox list, when bbox data is unavailable.
* Add tests with mocked/fake detector outputs so CI does not depend on downloading heavy detector weights.
* Coordinate the payload shape with frontend rendering tasks.
* Document the mapping, bbox contract, and limitations.

### Definition of Done

* [x] Implementation completed
* [x] Pretrained object detector integration or detector abstraction exists
* [x] Detector output can be represented as bbox candidates with object label, confidence, and coordinates
* [x] Mapping from `ActionEvent.label` to allowed object classes is implemented
* [x] Only bbox candidates relevant to the predicted action label are attached to the event payload
* [x] Multiple bboxes per action event are supported
* [x] Events without bboxes remain valid and backward-compatible
* [x] Bboxes use the contract from #115, including `box_format`, `coordinate_space`, `frame_index`, `source_width`, `source_height`, `x_min`, `y_min`, `x_max`, `y_max`, `label`, and `confidence`
* [x] Bounding box enrichment is integrated into the shared event pipeline or a compatible event enrichment hook
* [x] Main dashboard no longer depends on `mockCoordinates.json` for overlay boxes
* [x] Missing bbox data does not crash inference or event emission
* [x] Unit tests cover action-label-to-object-class mapping
* [x] Unit tests cover attaching filtered bboxes to `ActionEvent`
* [x] Tests use mocked/fake detector outputs and do not require downloading detector weights in CI
* [ ] Frontend team confirms the payload is renderable
* [x] Basic local verification completed on a sample video
* [x] Docs/README updated if relevant
* [x] Documentation clearly states that this is event-related object overlay, not full spatio-temporal action localization
* [x] PR created and linked to this issue

### PR DESCRIPTION
**Pull request by bartoszir**
## Summary
Implements object bounding-box generation for action events (issue #119). Detects objects on a representative frame from each inference window using a pretrained YOLO model, filters detections to only the object classes relevant to the predicted action label, and attaches the result to `ActionEvent.bboxes` using the contract from #115. Also fixes the `bbox_hook` contract in the shared pipeline (#127) so the hook has access to the raw frame window, not just event metadata.

**What changed:**

- **[CHANGED] [`src/inference/pipeline.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/src/inference/pipeline.py)** — `BBoxHook` changed from `Callable[[ActionEvent], ActionEvent]` to `Callable[[ActionEvent, InferenceResult], ActionEvent]`. `_run_bbox_hook` now forwards `InferenceResult` so hooks can access raw frame pixels.
- **[CHANGED] [`tests/inference/test_pipeline.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/tests/inference/test_pipeline.py)** — Updated the three hook signatures in `TestBBoxHook` to match the new contract. No assertion logic changed.
- **[NEW] [`src/inference/bbox_detector.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/src/inference/bbox_detector.py)** — `ACTION_LABEL_TO_OBJECT_CLASSES` mapping, `RawDetection` dataclass, `ObjectDetector` protocol, `YoloObjectDetector` (lazy `ultralytics` import, mirrors existing `torchvision` pattern in `context_policy.py`), and `BBoxEnricher` — the production component implementing `BBoxHook`: selects a representative frame, runs detection, filters to action-relevant classes, returns `event.model_copy(update={"bboxes": [...]})`. No relevant detections → `bboxes=None`, never an empty list.
- **[NEW] [`tests/inference/test_bbox_detector.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/tests/inference/test_bbox_detector.py)** — 16 tests covering label mapping, detection filtering, multiple bboxes per event, `BoundingBox` field correctness, frame-selector modes, empty-window safety, pipeline integration, and `YoloObjectDetector` validation — all using a fake detector, no real YOLO weights downloaded.
- **[CHANGED] [`requirements.txt`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/requirements.txt)** — Added `ultralytics` as an optional dependency. Not required to run the test suite.
- **[NEW] [`docs/backend.md`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/docs/backend.md)** — Documents the detection pipeline, label-to-class mapping table, integration point, and explicit limitations (event-related object overlay, not spatio-temporal action localization).
- **[CHANGED] [`.gitignore`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/.gitignore)** — Added `*.pt`/`*.pth`/`*.onnx` so downloaded model weights are never committed.

**Ownership boundary:** `InferenceEventPipeline` calls the hook and catches exceptions but knows nothing about YOLO. `BBoxEnricher` implements the hook contract but knows nothing about alert state or `EventPayload`. `ObjectDetector`/`YoloObjectDetector` perform pure object detection and know nothing about `ActionEvent` or label mapping.

**Out of scope (intentional):** This PR does not modify `ActionEvent`, `BoundingBox`, `ContextPolicy`, `AlertStateMachine`, or `EventPayload`. Frontend overlay rendering and removal of `mockCoordinates.json` are coordinated separately with the frontend team.

## Testing
- verified with:
  - `python -m pytest tests/inference/test_bbox_detector.py -v` — 16 passed
  - `python -m pytest tests/inference -q` — all passed
  - `python -m ruff check src/inference/bbox_detector.py src/inference/pipeline.py tests/inference/test_bbox_detector.py tests/inference/test_pipeline.py` — all checks passed
- manually verified against real footage in `data/raw/car_drops_off_person/` using `YoloObjectDetector` (not the fake detector) — confirmed correct `person`/`car` detections, sensible coordinates, and correct label filtering

## Linked Issue
Partially addresses #119

## Checklist
- [x] Changes were tested locally
- [x] CI passes
- [x] README/docs updated if relevant
- [x] Scope matches the linked Issue
- [x] No direct work outside the agreed scope

### CODE REVIEW

**Review by lukaszjecek — CHANGES_REQUESTED** [review]
Main blockers:

1. `BBoxEnricher` / `YoloObjectDetector` are not wired into the production MP4/RTSP or browser-camera flows. `run_inference()` and `CameraStreamSession` still construct the pipeline without `bbox_hook=`, so normal emitted events will still have `bboxes=None` unless a caller manually injects the hook.

2. The PR claims `Closes #119`, but issue #119 also requires the main dashboard to stop depending on mock bbox coordinates and frontend confirmation that the payload is renderable. The dashboard still uses mock bbox data, and frontend bbox consumption/rendering is still tracked separately in #131.

3. The tests prove the bbox hook works when manually injected, but they do not prove that real production MP4/RTSP/browser-camera events receive backend-generated bboxes.

Please either:
- wire bbox enrichment into the production runtime paths, add production-flow tests, and complete/verify the frontend payload path, or
- change this PR from `Closes #119` to `Partially addresses #119` and leave the remaining runtime/frontend integration work open.

One additional concern: `ultralytics` is added as an unconditional, unpinned root dependency, even though this is described as optional. Since `YOLO("yolov8n.pt")` may download weights at runtime, the PR should define reproducible deployment behavior: pinned dependency, configured local weights path, detector initialization failure fallback, and detector reuse/caching.

**Comment by bartoszir** [comment]
## Update: addressing review feedback

This PR previously only exposed `BBoxEnricher`/`YoloObjectDetector` for manual injection. The following changes wire it into production and make the YOLO dependency reproducible, per review:

**What changed (added on top of the above):**

- **[CHANGED] [`src/inference/bbox_detector.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/src/inference/bbox_detector.py)** — `YoloObjectDetector` gains a `weights_dir` parameter to load weights from a local path instead of downloading at runtime (`FileNotFoundError` if missing). Added `get_or_create_bbox_enricher()`, a thread-safe cache (mirrors the existing `ModelCache` pattern) so YOLO is not reloaded on every camera reconnect or new offline session.
- **[CHANGED] [`requirements.txt`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/requirements.txt)** — Pinned `ultralytics==8.4.75` instead of an unpinned dependency.
- **[CHANGED] [`src/inference/runtime.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/src/inference/runtime.py)** — `InferenceRuntimeSettings` gains `bbox_enabled` (default `False`, explicit opt-in), `bbox_model_name`, `bbox_confidence_threshold`, `bbox_weights_dir`, `bbox_frame_selector`, with validation.
- **[CHANGED] [`configs/data_pipeline.yml`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/configs/data_pipeline.yml)** — Added a `bbox:` section mirroring the new settings fields.
- **[CHANGED] [`src/inference/service.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/src/inference/service.py)** — `run_inference()` now initializes `bbox_hook` in a `try/except` block identical in style to the existing `context_module` initialization, and passes it to `InferenceEventPipeline`. Initialization failure (missing `ultralytics`, missing weights, network error) logs a warning and falls back to `bbox_hook=None` — never crashes the pipeline.
- **[CHANGED] [`src/app/services/camera_stream_manager.py`](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/blob/119-feature-add-object-bbox-generation-and-event-to-object-mapping-for-overlay-payload/src/app/services/camera_stream_manager.py)** — `CameraStreamSession.__init__` follows the same pattern (`self.bbox_hook`), wiring browser-camera sessions into the same opt-in bbox flow as MP4/RTSP.
- **[NEW]** 10 new tests across `tests/inference/test_bbox_detector.py` (4), `tests/inference/test_service.py` (3), and `tests/app/test_camera_stream_manager.py` (3) — covering `weights_dir` resolution and caching, and confirming both production entry points wire the hook when enabled, leave it `None` by default, and fall back to `None` on initialization failure. All YOLO calls are mocked; no real weights are downloaded in CI.

**Addressed from review:**
1. Bbox enrichment is now wired into both production runtime paths (MP4/RTSP via `run_inference()`, browser-camera via `CameraStreamSession`), behind an explicit `bbox_enabled` flag (default `False`, so no existing test or deployment is affected).
2. `ultralytics` is now pinned, supports a local `weights_dir` for reproducible offline deployment, is cached/reused across sessions, and fails closed (falls back to no bboxes) rather than crashing inference.
3. New tests prove both production entry points actually wire the hook end-to-end, not just that the hook works when manually injected.

**Still out of scope (tracked in #131):** Removing `mockCoordinates.json` from the dashboard and frontend confirmation that the payload is renderable. Kept as `Partially addresses #119` until that work lands.

## Testing (updated)
- verified with:
  - `python -m pytest tests/inference tests/app -q` — 354 passed
  - `python -m ruff check` on all changed/added files — all checks passed

**Review by lukaszjecek — APPROVED** [review]
APPROVE

Backend scope looks good now: bbox enrichment is wired into MP4/RTSP and browser-camera paths, and the PR is correctly marked as `Partially addresses #119`.

@Filipwasiel please verify the remaining frontend scope around bbox rendering / `mockCoordinates.json` and either approve from your side or leave any blockers.

**Review by Filipwasiel — APPROVED** [review]

**Review by alicjaswiers — APPROVED** [review]

**Review by Szymon110903 — APPROVED** [review]
